ctapipe/image/extractor.py

```diff
@@ -1285,3 +1289,400 @@ def __call__(
+        Shape: (n_pix, upsampling * n_samples)
+    """
+    deconvolved_waveforms = np.atleast_2d(waveforms) - np.atleast_2d(baselines).T
+    deconvolved_waveforms[:, 1:] -= pole_zero * deconvolved_waveforms[:, :-1]
+    deconvolved_waveforms[:, 0] = 0
+
+    def filtfilt_custom(signal, filt):
+        forward = convolve1d(signal, filt, axis=-1, mode="nearest")
+        backward = convolve1d(forward[::-1], filt, axis=-1, mode="nearest")
+        return backward[::-1]
+
+    if upsampling > 1:
+        filt = np.ones(upsampling)
+        filt_weighted = filt / upsampling
+        signal = np.repeat(deconvolved_waveforms, upsampling, axis=-1)
+        return filtfilt_custom(signal, filt_weighted)
+
+    return deconvolved_waveforms
+
+
+@guvectorize(
+    [
+        (float64[:], int64, float64, float32[:]),
+        (float32[:], int64, float64, float32[:]),
+    ],
+    "(s),(),()->()",
+    nopython=True,
+    cache=True,
+)
+def adaptive_centroid(waveforms, peak_index, rel_descend_limit, centroids):
+    """
+    Calculates the pulse centroid for all samples down to rel_descend_limit * peak_amplitude.
+
+    The ret argument is required by numpy to create the numpy array which is
+    returned. It can be ignored when calling this function.
+
+    Parameters
+    ----------
+    waveforms : ndarray
+        Waveforms stored in a numpy array.
+        Shape: (n_pix, n_samples)
+    peak_index : ndarray or int
+        Peak index for each pixel.
+    rel_descend_limit : ndarray or float
+        Fraction of the peak value down to which samples are accumulated in the centroid calculation.
+    centroids : ndarray
+        Return argument for ufunc (ignore)
+        Returns the peak centroid in units "samples"
+
+    Returns
+    -------
+    centroids : ndarray
+        Peak centroid in units "samples"
+    """
+    centroids[0] = peak_index  # preload in case of errors
+
+    n_samples = waveforms.size
+    if n_samples == 0:
+        return
+
+    if (peak_index > (n_samples - 1)) or (peak_index < 0):
+        raise ValueError("peak_index must be within the waveform limits")
+
+    peak_amplitude = waveforms[peak_index]
+    if peak_amplitude < 0.0:
+        return
+
+    descend_limit = rel_descend_limit * peak_amplitude
+
+    sum_ = float64(0.0)
+    jsum = float64(0.0)
+
+    j = peak_index
+    while j >= 0 and waveforms[j] > descend_limit:
+        sum_ += waveforms[j]
+        jsum += j * waveforms[j]
+        j -= 1
+        if waveforms[j] > peak_amplitude:
+            descend_limit = rel_descend_limit * peak_amplitude
+
+    j = peak_index + 1
+    while j < n_samples and waveforms[j] > descend_limit:
+        sum_ += waveforms[j]
+        jsum += j * waveforms[j]
+        j += 1
+        if waveforms[j] > peak_amplitude:
+            descend_limit = rel_descend_limit * peak_amplitude
+
+    if sum_ != 0.0:
+        centroids[0] = jsum / sum_
+
+
+class FlashCamExtractor(ImageExtractor):
+    """
```

Comment:
This docstring is missing a small header line like "Image extractor applying signal preprocessing for FlashCam"